chained/algeneva_impl.go

```diff
@@ -20,13 +24,34 @@ type algenevaImpl struct {
 }
 
 func newAlgenevaImpl(addr string, pc *config.ProxyConfig, reportDialCore reportDialCoreFn) (*algenevaImpl, error) {
-	strategy := ptSetting(pc, "algeneva_strategy")
+	opts := algeneva.DialerOpts{
+		AlgenevaStrategy: ptSetting(pc, "algeneva_strategy"),
+	}
+
+	if cert := pc.Cert; cert != "" {
+		block, rest := pem.Decode([]byte(pc.Cert))
+		if block == nil {
+			return nil, errors.New("failed to decode proxy certificate as PEM block")
+		}
+
+		if len(rest) > 0 {
+			return nil, errors.New("unexpected extra data in proxy certificate PEM")
+		}
 
-	ops := algeneva.DialerOpts{
-		AlgenevaStrategy: strategy,
+		if block.Type != "CERTIFICATE" {
+			return nil, errors.New("expected certificate in PEM block")
+		}
+
+		certPool := x509.NewCertPool()
+		certPool.AppendCertsFromPEM([]byte(cert))
+		opts.TLSConfig = &tls.Config{
+			RootCAs:    certPool,
+			ServerName: pc.TLSServerNameIndicator,
+		}
```

Comment:
This would normally be the correct way to set this up.  However, the proxy cert is not signed for the hostname configured in `pc.TLSServerNameIndicator`.  That field is used to set a decoy value to throw off passive observers since the TLS server name indicator is transmitted in the clear.  I think you could just set this to the public IP of the proxy (which you can grab by parsing `addr`).

For reference, [we usually pin the proxy's certificate](https://github.com/getlantern/flashlight/blob/b36afb1d310a894232c26a31adefbc66eb419216/chained/tls_config.go#L74-L114).  We do this because Go doesn't offer a mechanism for both (1) advertising the SNI we want and (2) ignoring the mismatch between the SNI and the hostname in the proxy's certificate.  In the case of `algeneva`, it is not important that we advertise the bogus SNI.